chemclipse/plugins/org.eclipse.chemclipse.xxd.model/src/org/eclipse/chemclipse/xxd/model/filter/peaks/ProcessPeaksByPeakClassifierFilter.java

```diff
@@ -0,0 +1,169 @@
+
+		return "Peak Classifier Filter";
+	}
+
+	@Override
+	public String getDescription() {
+
+		return "Filter peaks and add peak classifier (e.g. for whole name or parts of the name.)";
+	}
+
+	@Override
+	public Class<ProcessPeaksByPeakClassifierFilterSettings> getConfigClass() {
+
+		return ProcessPeaksByPeakClassifierFilterSettings.class;
+	}
+
+	@Override
+	public boolean acceptsIPeaks(Collection<? extends IPeak> items) {
+
+		return true;
+	}
+
+	@Override
+	public <X extends IPeak> void filterIPeaks(CRUDListener<X, IPeakModel> listener, ProcessPeaksByPeakClassifierFilterSettings configuration, MessageConsumer messageConsumer, IProgressMonitor monitor) throws IllegalArgumentException {
+
+		Collection<X> read = listener.read();
+		if(configuration == null) {
+			configuration = createConfiguration(read);
+		}
+		SubMonitor subMonitor = SubMonitor.convert(monitor, read.size());
+		LinkedHashMap<String, String> parsedUserDefinedValues = parseUserDefinedValuesAsMap(configuration);
+
+		for(X peak : read) {
+			setPeakClassifier(configuration, parsedUserDefinedValues, peak);
+			subMonitor.worked(1);
+		}
+	}
+
+	private static LinkedHashMap<String, String> parseUserDefinedValuesAsMap(ProcessPeaksByPeakClassifierFilterSettings configuration) {
+
+		String userDefinedMatchExpression = configuration.getUserDefinedMatchExpression();
+		String matchClassification = configuration.getMatchClassification();
+		List<List<String>> matchList = new ArrayList<List<String>>();
+		if(!userDefinedMatchExpression.isEmpty()) {
+			parseExpressions(matchList, userDefinedMatchExpression);
+		} else {
+			throw new IllegalArgumentException("No match expressions defined!");
+		}
+		if(!matchClassification.isEmpty()) {
+			parseExpressions(matchList, matchClassification);
+		} else {
+			throw new IllegalArgumentException("No match classifications defined!");
+		}
+		if(matchList.get(0).size() != matchList.get(1).size()) {
+			throw new IllegalArgumentException("The number of Expressions is not the same as the number of Classifications.");
+		}
+		LinkedHashMap<String, String> parsedUserDefinedValues = new LinkedHashMap<String, String>();
+		for(int i = 0; i < matchList.get(0).size(); i++) {
+			parsedUserDefinedValues.put(matchList.get(0).get(i),matchList.get(1).get(i));
+		}
+		return parsedUserDefinedValues;
+	}
+
+	private static void parseExpressions(List<List<String>> matchList, String userExpressions) {
+
+		List<String> expressionsList = new ArrayList<String>();
+		String[] tempExpressions = userExpressions.split(System.getProperty("line.separator"));
+		for(int i = 0; i < tempExpressions.length; i++) {
+			if(!tempExpressions[i].isEmpty()) {
+				expressionsList.add(tempExpressions[i]);
+			}
+		}
+		matchList.add(expressionsList);
+	}
+
+	private static <X extends IPeak> void setPeakClassifier(ProcessPeaksByPeakClassifierFilterSettings configuration, LinkedHashMap<String, String> parsedUserDefinedValues, X peak) {
+
+		peak.removeClassifier("");
```

Comment:
What is this for?

Reply:
This is to remove an empty String that is always present as classifier in the model.

So the added Classifier are displayed in the peak list like "Classifier 1, Classifier 2," instead of ", Classifier 1, Classifier 2"

Reply:
OK, then the peak classifier model should be improved. Can you open a new issue for that?

Reply:
Please keep in mind that there are two types of classifications atm: Peak Classifications and Target classifications.
If this is really for classifying substances IMO the detected target substance should be classified not the peak.

Reply:
As Philip said, the current implementation peak.addClassifier() is correct. The purpose is to add a sorting property in the UI/peak/scan list .